packages/pyright-internal/src/analyzer/typeNameUtils.ts

```diff
@@ -0,0 +1,90 @@
+
+    // If there are no full matches, try to find types whose name is a prefix to `type`’s full name
+    // and then recursively try to find a symbol with `type`’s full type.
+    return symbols
+        .filter((symbolInfo) => fullName.startsWith(symbolInfo.fullName))
+        .flatMap((symbolInfo) =>
+            getLocalTypeNameRec(evaluator, type, symbolInfo.type).map((name) => `${symbolInfo.localName}.${name}`)
+        );
+}
+
+function getLocalTypeNameRec(
+    evaluator: TypeEvaluator,
+    type: ClassType | ModuleType,
+    symbolType: ClassType | ModuleType
+) {
+    return handleSymbolTable(
+        evaluator,
+        type,
+        isClass(symbolType) ? ClassType.getSymbolTable(symbolType) : symbolType.priv.fields
+    );
+}
+
+/**
+ * Determine the names (potentially including `.`) which can be used to reference `type` within `scope`.
+ *
+ * This function goes through the symbols defined within `scope` and its parent to find classes/modules
+ * whose full name is a prefix of `type`’s full name. For each of these, it recursively searches the symbols
+ * defined in that class/module for other classes/modules whose full name is a prefix of `type`’s full name
+ * until a type with `type`’s full name is found.
+ *
+ * For example, consider the following definition:
+ * ```python
+ * class A:
+ *     class B: pass
+ * ```
+ * After this, the type representing `A.B` has the local names `['A.B']`. However, if that definition is
+ * in a different module called `lib`, we have a few different cases for the local names of `A.B`
+ * depending on how `lib`/its members have been imported:
+ * - `import lib`: The local names are `['lib.A.B']`.
+ * - `import lib as mod`: The local names are `['mod.A.B']`.
+ * - `import lib as mod; from lib import A`: The local names are `['mod.A.B', 'A.B']`.
+ * - `import lib as mod; from lib import A as C`: The local names are `['mod.A.B', 'C.B']`.
+ */
+export function getLocalTypeNames(evaluator: TypeEvaluator, type: ClassType | ModuleType, scope: Scope): string[] {
+    const out: string[] = [];
+    let currentScope: Scope | undefined = scope;
+    while (currentScope) {
+        out.push(...handleSymbolTable(evaluator, type, currentScope.symbolTable));
+        currentScope = currentScope.parent;
+    }
+    return out;
+}
+
+/**
+ * Determine the parts names of `type` and its enclosing classes from least to most nested class.
+ *
+ * For example, if class `C` is defined in class `B`, which in turn is defined in class `A`,
+ * this function returns `['A', 'B', 'C']`.
+ */
+export function getNestedClassNameParts(type: ClassType): string[] {
+    const parts: string[] = [type.shared.name];
+    let enclosingNode = type.shared.declaration ? getEnclosingClass(type.shared.declaration?.node, false) : undefined;
+    while (enclosingNode) {
+        parts.push(enclosingNode.d.name.d.value);
+        enclosingNode = getEnclosingClass(enclosingNode, false);
+    }
+    return parts.reverse();
+}
```

Comment:
`type.shared.fullName` already contains this information so i don't think we need to traverse the nodes like this. can just do something like `type.shared.fullName.split('.').slice(1)`